Property docs

karlbom opened the pull request:
## Summary

This PR is going to add API documentation as well es introductory documentation on how to interact and with and use `Properties`, `Property DDS` and `Property Binder`(aka `DataBinder`)

At the moment the API documentation lives under the experimental section in the docs section:

![image](https://user-images.githubusercontent.com/43415869/117819272-adac6a00-b269-11eb-8df2-46c05c826aa5.png)

Additional Documentation and Guides I currently put in the getting started page in its own `Property` section.

![image](https://user-images.githubusercontent.com/43415869/117819518-e8ae9d80-b269-11eb-9630-386c21a45714.png)


I marked this PR as draft as the work on the documentation is still ongoing but it is probably a good idea to share the current state to discuss where these docs should be located and anything related to it.

There are some items still need to be done:
## Open Items
- [x] There are some changes I needed to do to the `custom-api-documenter` @tylerbutler provides a possible solution but there is still a bugfix that needs to go into the package (see https://github.com/karlbom/custom-api-documenter/commit/2918651ad07e949ba9e4059ae084285e1964be3b)
- [ ] final location of API docs
- [ ] final location of documentation and user guides etc.
- [x]  add Property DDS documentation
- [x]  add Property documentation (create, modify etc.)

Marking @SamBroner @DLehenbauer

karlbom committed:
modify configs to include property documentation

karlbom committed:
adopt workaround fork of custom-api-documenter

karlbom committed:
wip

karlbom committed:
update documentation

karlbom committed:
doc update

karlbom committed:
progress

karlbom committed:
tweak binder API documentation

karlbom committed:
update description on Property DDS classes

DLehenbauer commented:
@karmbom - FYI - Running `npm run policy-check:fix` from the root of the enlistment will sort the package.json file for you.
@tylerbutler - Thanks for helping out!

karlbom committed:
fix script ordering

karlbom commented:
@tylerbutler  I made a PR to your branch incorporating the changes I made to the custom-api-documenter  ( https://github.com/tylerbutler/custom-api-documenter/pull/1)

karlbom committed:
split dds and binder documentation
change property binder/binding to databinding
to stay consistend with API documentation

karlbom committed:
add property factory description

tylerbutler commented:
@karlbom Thank you! I was hoping to switch to [this new fork](https://github.com/tylerbutler/rushstack/tree/custom-api-documenter) but when I tested yesterday the output wasn't what I expected, so I need to debug more. I don't want to hold things up, so I'll merge [your PR](https://github.com/tylerbutler/custom-api-documenter/pull/1) and update main to use it.

The fixes you made should already be included in my new fork, with the exception of the backwards-compatible schema change.

karlbom committed:
add descrption for property types

karlbom committed:
update overview

karlbom commented:
Since the documentation turned out to be quite a lot I decided for the moment to split them into handling the basics of Property DDS and properties and the documentation around the Property DataBinder

![image](https://user-images.githubusercontent.com/43415869/118265123-06be0d00-b4b9-11eb-8b7b-285975fdfc9f.png)


@SamBroner you mentioned there is going to be some update around the documentation do you have any opinion on where the documentation for this PR should live or whether it is okay to be where it is for now?

karlbom committed:
adding basic propertys api

karlbom committed:
change all relevant exports to generate documentation

karlbom committed:
fix some documentation
add custom entrypoint for property-properties

karlbom committed:
add missing header

karlbom committed:
fixed imports that were overlooked

karlbom committed:
fix experimental change

karlbom committed:
cleanup

karlbom committed:
fix imports in tests

tylerbutler commented:
I'd like to propose doing one of two things:

1. Create a new section for DDSes and add the Property DDS docs there. OR
2. Create new empty sections under `docs` for all the DDSes, and add the Property DDS docs there.

Option 2 is probably the most expedient, though I worry about the nav being too complex. @SamBroner Did you have something else in mind?

SamBroner commented:
> I'd like to propose doing one of two things:
> 
> 1. Create a new section for DDSes and add the Property DDS docs there. OR
> 2. Create new empty sections under `docs` for all the DDSes, and add the Property DDS docs there.
> 
> Option 2 is probably the most expedient, though I worry about the nav being too complex. Sam Broner Did you have something else in mind?

Let's start with 2, that is what I had in mind. I should have a PR for that tomorrow afternoon.

tylerbutler commented:
Sounds good. Let me know if you need help.

karlbom commented:
@SamBroner is there an update?

Commit:
initial spelling & formating fixes

karlbom commented:
@SamBroner thanks for the fixes

karlbom committed:
Merge branch 'main' into property_docs

kailui commented on docs/content/start/property-dds/property-overview.md:
```diff
@@ -0,0 +1,96 @@
+### What is Property Factory?
+
+The is a factory object for instantiating Properties. It comes with a set of rich primitive types (Float32, Enum, Uint16
+, Uint32, etc. and custom types) For a more detailed explanation please refer to the
```
```suggestion
, Uint32, etc. and custom types). For a more detailed explanation please refer to the
```

DLehenbauer replied:
@kailui - Thank you for contributing proof reading fixes!

kailui commented on docs/content/start/property-dds/property-overview.md:
```diff
@@ -0,0 +1,96 @@
+A Commit is the “unit of change” in Property DDS. It stores an atomic transaction recording every insert, modify and
+delete operations that happened since the previous commit.
+
+Much like in Git, a chain of commits is known as a “Branch”. The state of the is constructed by walking along the history and applying the commits on top of the previous one at any point in time. It is important to note that — since history cannot be modified — a must be referring the most recent state in order to persist any changes in Property DDS.
```
Sounds like a word is missing here.

kailui replied:
> a must be referring

Should there be a word after "a"?

kailui commented on docs/content/start/property-dds/property-overview.md:
```diff
@@ -0,0 +1,96 @@
+A Commit is the “unit of change” in Property DDS. It stores an atomic transaction recording every insert, modify and
+delete operations that happened since the previous commit.
+
+Much like in Git, a chain of commits is known as a “Branch”. The state of the is constructed by walking along the history and applying the commits on top of the previous one at any point in time. It is important to note that — since history cannot be modified — a must be referring the most recent state in order to persist any changes in Property DDS.
```
> The state of the is constructed by

Sounds like a word is missing between "the" and "is".

karlbom replied:
thanks I added the missing words in that section

kailui commented on docs/content/start/property-dds/property-overview.md:
```diff
@@ -0,0 +1,96 @@
+
+## Read Data From The PropertySet
+
+Naturally, after writing data to a Fluid Property DDS we need the ability to retrieve this data. For this, we must initialize the Property DDS with the Fluid `documentId`from the Property DDS that we wrote or data into. Once the *Property DDS* is fully initialized by the FluidFramework you can retrieve that data.
```
```suggestion
Naturally, after writing data to a Fluid Property DDS we need the ability to retrieve this data. For this, we must initialize the Property DDS with the Fluid `documentId`from the Property DDS that we wrote our data into. Once the Property DDS is fully initialized by the Fluid Framework you can retrieve that data.
```

kailui commented on docs/content/start/property-dds/property-schemas.md:
```diff
@@ -0,0 +1,320 @@
+A *Property Sets Schema* defines the structure of a Property Set. It defines which child properties exist in a
+property and which types they have.
+
+The following example shows a possible schema for your rectangle. This example is only meant to show the various possibilities in defining schemas and is not necessarily showing the best way to define a rectangle.
```
```suggestion
The following example shows a possible schema for your rectangle. This example is only meant to show the various possibilities in defining schemas and is not necessarily the best way to define a rectangle.
```

kailui commented on docs/content/start/property-dds/property-schemas.md:
````diff
@@ -0,0 +1,320 @@
+}
+```
+
+Or you can refer to an external schema, as we did for fillcolor.
````
```suggestion
Or you can refer to an external schema, as we did for fill color.
```

kailui commented on docs/content/start/property-dds/property-schemas.md:
```diff
@@ -0,0 +1,320 @@
+
+
+
+## nnotation
```
```suggestion
## Annotation
```

kailui commented on docs/content/start/property-dds/property-schemas.md:
```diff
@@ -0,0 +1,320 @@
+
+    Each constant should have an id, value and a valid typeid. The id must be unique per declared constants and properties, that is a constant and a property cannot have the same id.
+
+    A constant can be of primitive, enum, string, array, set, map, or custom type.
```
```suggestion
A constant can be of primitive, enum, string, array, set, map, or custom type.
```

kailui commented on docs/content/start/property-dds/property-schemas.md:
```diff
@@ -0,0 +1,320 @@
+
+    A constant can be of primitive, enum, string, array, set, map, or custom type.
+
+    They are read only properties; values and instances can be retrieved like any other properties.
```
```suggestion
They are read-only properties; values and instances can be retrieved like any other properties.
```

kailui commented on docs/content/start/property-dds/property-schemas.md:
```diff
@@ -0,0 +1,320 @@
+    A constant can be of primitive, enum, string, array, set, map, or custom type.
+
+    They are read only properties; values and instances can be retrieved like any other properties.
+    Constants can be overridden in inherited schemas. To do that, the inherited constant should keep the same 'id', 'typeid' and 'context' as the base one it overrides.
```
```suggestion
Constants can be overridden in inherited schemas. To do that, the inherited constant should keep the same 'id', 'typeid' and 'context' as the base one it overrides.
```

kailui commented on docs/content/start/property-dds/property-primitive.md:
```diff
@@ -0,0 +1,72 @@
+| **Float32**   |   A 32 bit floating point value (according to IEEE 754) |
+| **Float64**   |  A 64 bit floating point value (according to IEEE 754) |
+| **Int8**      | A signed 8 bit integer value (with a valid range from −128 to 127) |
+| **Uint8**     | A unsigned 8 bit integer value (with a valid range from 0 to 255) |
```
```suggestion
| **Uint8**     | An unsigned 8 bit integer value (with a valid range from 0 to 255) |
```

kailui commented on docs/content/start/property-dds/property-primitive.md:
```diff
@@ -0,0 +1,72 @@
+| **Int8**      | A signed 8 bit integer value (with a valid range from −128 to 127) |
+| **Uint8**     | A unsigned 8 bit integer value (with a valid range from 0 to 255) |
+| **Int16**     | A signed 16 bit integer value (with a valid range from −32768 to 32767) |
+| **Uint16**    | A unsigned 16 bit integer value (with a valid range from 0 to 65536) |
```
```suggestion
| **Uint16**    | An unsigned 16 bit integer value (with a valid range from 0 to 65536) |
```

kailui replied:
```suggestion
| **Uint16**    | An unsigned 16 bit integer value (with a valid range from 0 to 65536) |
```

kailui commented on docs/content/start/property-dds/property-primitive.md:
```diff
@@ -0,0 +1,72 @@
+| **Int64**     | A signed 64 bit integer value (with a valid range from -9,223,372,036,854,775,808 to 9,223,372,036,854,775,807) |
+| **Uint64**    | An unsigned 64 bit integer value (with a valid range from 0 to 18,446,744,073,709,551,615) |
+| **Bool**      | A boolean value, either true or false |
+| **String**    | An unicode string |
```
```suggestion
| **String**    | A Unicode string |
```

kailui commented on docs/content/start/property-dds/property-paths-references.md:
```diff
@@ -0,0 +1,451 @@
+
+// Now we can resolve paths
+root.resolvePath('point.x');
+root.resolvePath('map[point].x');
```
```suggestion
root.resolvePath('map[point2].x');
```

kailui commented on docs/content/start/property-dds/property-paths-references.md:
````diff
@@ -0,0 +1,451 @@
+root.resolvePath(PropertyUtils.PathHelper.quotePathSegmentIfNeeded('"doubleQuotes"'));
+```
+
+A path that starts with a ``/`` it is an **absolute path**. This means the resolution will start at the root of the
````
```suggestion
A path that starts with a ``/`` is an **absolute path**. This means the resolution will start at the root of the
```

kailui commented on docs/content/start/property-dds/property-collection.md:
```diff
@@ -0,0 +1,291 @@
+**Primitive Collections**
+  Primitive collections are strictly typed with one of the Primitive Property Types (e.g. a String array or Float64
+  map). These collections directly contain the literal values of the given type. They do not support polymorphism, all
+  entries have to be of the same type The collection itself is a leaf in the PropertySet   tree. This means, there are
```
```suggestion
  entries have to be of the same type. The collection itself is a leaf in the PropertySet tree. This means, there are
```

karlbom committed:
Update docs/content/start/property-dds/property-collection.md

Co-authored-by: kailui <kailui@microsoft.com>

karlbom committed:
Update docs/content/start/property-dds/property-paths-references.md

Co-authored-by: kailui <kailui@microsoft.com>

karlbom committed:
Update docs/content/start/property-dds/property-paths-references.md

Co-authored-by: kailui <kailui@microsoft.com>

karlbom committed:
Apply suggestions from code review

Co-authored-by: kailui <kailui@microsoft.com>

karlbom committed:
Apply suggestions from code review (2)

Co-authored-by: kailui <kailui@microsoft.com>

karlbom committed:
fix missing words in overview

tylerbutler commented:
I'm adding an experimental section back to the docs in an upcoming PR. I'll include these docs in that section.
